src/core/ddsc/src/dds_write.c

```diff
@@ -270,19 +271,147 @@ static dds_return_t deliver_locally (struct writer *wr, struct ddsi_serdata *pay
   return rc;
 }
 
```

Comment:
General note: we can improve some of the function names, bool variable condition names and potentially optimize the code path in dds_write_impl even more (wrt. branching and condition evaluation), but this was secondary to eliminate the serialization and general consolidation of the different ways of writing.

There is still a lot of room for improvement and refactoring.

Reply:
Refactored conditions somewhat, volatile writer check is now an assert as per discussion with @eboasson